File: backend/models/task.py
"""Task models for FastAPI Email Helper API."""

from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field
from enum import Enum


class TaskStatus(str, Enum):
    """Task status enumeration."""
    TODO = "todo"
    PENDING = "pending"
    IN_PROGRESS = "in-progress" 
    REVIEW = "review"
    DONE = "done"
    COMPLETED = "completed"
    CANCELLED = "cancelled"


class TaskPriority(str, Enum):
    """Task priority enumeration."""
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    URGENT = "urgent"


class TaskBase(BaseModel):
    """Base task model."""
    title: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = None
    status: TaskStatus = TaskStatus.TODO
    priority: TaskPriority = TaskPriority.MEDIUM
    category: Optional[str] = None
    due_date: Optional[datetime] = None
    tags: Optional[list[str]] = None
    metadata: Optional[dict] = None


class TaskCreate(TaskBase):
    """Task creation model."""
    email_id: Optional[str] = None


class TaskUpdate(BaseModel):
    """Task update model."""
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    status: Optional[TaskStatus] = None
    priority: Optional[TaskPriority] = None
    due_date: Optional[datetime] = None
    email_id: Optional[str] = None


class TaskListResponse(BaseModel):
    """Response model for paginated task lists."""
    tasks: list["Task"]
    total_count: int
    page: int
    page_size: int
    has_next: bool


class BulkTaskUpdate(BaseModel):
    """Model for bulk task updates."""
    task_ids: list[int]
    updates: TaskUpdate


class BulkTaskDelete(BaseModel):
    """Model for bulk task deletion."""
    task_ids: list[int]


class TaskInDB(TaskBase):
    """Task model as stored in database."""
    id: int
    created_at: datetime
    updated_at: datetime
    email_id: Optional[str] = None
    user_id: Optional[int] = None

    model_config = {"from_attributes": True}


class Task(TaskBase):
    """Task model for API responses."""
    id: str  # String for frontend compatibility
    created_at: datetime
    updated_at: datetime
    email_id: Optional[str] = None

    model_config = {"from_attributes": True}

File: backend-go/internal/models/task.go
package models

import "time"

// Task represents a task/action item
type Task struct {
	ID             int        `json:"id" db:"id"`
	UserID         int        `json:"user_id" db:"user_id"`
	Title          string     `json:"title" db:"title"`
	Description    string     `json:"description,omitempty" db:"description"`
	Status         string     `json:"status" db:"status"` // pending, in_progress, completed, cancelled
	Priority       string     `json:"priority" db:"priority"` // low, medium, high, urgent
	Category       string     `json:"category,omitempty" db:"category"`
	EmailID        *string    `json:"email_id,omitempty" db:"email_id"`
	OneLineSummary *string    `json:"one_line_summary,omitempty" db:"one_line_summary"`
	DueDate        *time.Time `json:"due_date,omitempty" db:"due_date"`
	CreatedAt      time.Time  `json:"created_at" db:"created_at"`
	UpdatedAt      time.Time  `json:"updated_at" db:"updated_at"`
	CompletedAt    *time.Time `json:"completed_at,omitempty" db:"completed_at"`
}

// TaskCreate request for creating a task
type TaskCreate struct {
	Title          string     `json:"title" binding:"required"`
	Description    string     `json:"description,omitempty"`
	Status         string     `json:"status,omitempty"`
	Priority       string     `json:"priority,omitempty"`
	Category       string     `json:"category,omitempty"`
	EmailID        *string    `json:"email_id,omitempty"`
	OneLineSummary *string    `json:"one_line_summary,omitempty"`
	DueDate        *time.Time `json:"due_date,omitempty"`
}

// TaskUpdate request for updating a task
type TaskUpdate struct {
	Title          *string    `json:"title,omitempty"`
	Description    *string    `json:"description,omitempty"`
	Status         *string    `json:"status,omitempty"`
	Priority       *string    `json:"priority,omitempty"`
	Category       *string    `json:"category,omitempty"`
	EmailID        *string    `json:"email_id,omitempty"`
	OneLineSummary *string    `json:"one_line_summary,omitempty"`
	DueDate        *time.Time `json:"due_date,omitempty"`
}

// TaskListResponse paginated task list
type TaskListResponse struct {
	Tasks      []Task `json:"tasks"`
	TotalCount int    `json:"total_count"`
	Page       int    `json:"page"`
	PageSize   int    `json:"page_size"`
	HasNext    bool   `json:"has_next"`
}

// BulkTaskUpdate for bulk updates
type BulkTaskUpdate struct {
	TaskIDs []int      `json:"task_ids" binding:"required"`
	Updates TaskUpdate `json:"updates" binding:"required"`
}

// BulkTaskDelete for bulk deletions
type BulkTaskDelete struct {
	TaskIDs []int `json:"task_ids" binding:"required"`
}
